File: internal/oauth/jwt.go
// Package oauth implements a minimal OAuth 2.1 Authorization Server + Bearer
// JWT resource-server middleware tailored to the edookit-mcp single-user
// remote-MCP deployment.
//
// Access tokens are HMAC-SHA256-signed JWTs (HS256). Symmetric keys are fine
// because the same process issues and validates the tokens — there is no
// separate verifier that would benefit from public-key verification.
package oauth

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

// jwtHeader is the fixed HS256 JWT header — emitted as-is, never parsed back.
// Keeping it as a literal makes the hot path allocation-free.
const jwtHeader = `{"alg":"HS256","typ":"JWT"}`

// jwtClaims is the access-token payload. Fields are the standard registered
// claims (RFC 7519) plus a free-form `scope` so the resource handler can see
// what the token was issued for.
type jwtClaims struct {
	Iss   string `json:"iss"`
	Sub   string `json:"sub"`
	Aud   string `json:"aud"`
	Iat   int64  `json:"iat"`
	Exp   int64  `json:"exp"`
	Jti   string `json:"jti"`
	Scope string `json:"scope,omitempty"`
}

// issueJWT produces a signed access token for the given subject and scope.
// Expiry is now + cfg.AccessTTL; iat/jti are filled in here.
func (s *Server) issueJWT(sub, scope string) (string, error) {
	now := s.now()
	jti, err := randomToken(16)
	if err != nil {
		return "", fmt.Errorf("jti: %w", err)
	}
	claims := jwtClaims{
		Iss:   s.cfg.PublicURL,
		Sub:   sub,
		Aud:   s.cfg.Audience,
		Iat:   now.Unix(),
		Exp:   now.Add(s.cfg.AccessTTL).Unix(),
		Jti:   jti,
		Scope: scope,
	}
	payload, err := json.Marshal(claims)
	if err != nil {
		return "", fmt.Errorf("marshal claims: %w", err)
	}
	headerB64 := base64.RawURLEncoding.EncodeToString([]byte(jwtHeader))
	payloadB64 := base64.RawURLEncoding.EncodeToString(payload)
	signing := headerB64 + "." + payloadB64
	mac := hmac.New(sha256.New, s.cfg.JWTSecret)
	mac.Write([]byte(signing))
	sigB64 := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	return signing + "." + sigB64, nil
}

// verifyJWT parses, signature-verifies and claim-checks a token issued by this
// server. Returns the subject ("sub" claim) on success. On any failure the
// caller should treat the request as anonymous — the error string is intended
// for the WWW-Authenticate error_description, not for debug logs.
func (s *Server) verifyJWT(tok string) (string, error) {
	parts := strings.Split(tok, ".")
	if len(parts) != 3 {
		return "", errors.New("malformed JWT")
	}
	headerB64, payloadB64, sigB64 := parts[0], parts[1], parts[2]

	// Constant-time signature compare via hmac.Equal: avoids timing oracle.
	mac := hmac.New(sha256.New, s.cfg.JWTSecret)
	mac.Write([]byte(headerB64 + "." + payloadB64))
	expectedSig := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	if !hmac.Equal([]byte(sigB64), []byte(expectedSig)) {
		return "", errors.New("bad signature")
	}

	payload, err := base64.RawURLEncoding.DecodeString(payloadB64)
	if err != nil {
		return "", fmt.Errorf("decode payload: %w", err)
	}
	var c jwtClaims
	if err := json.Unmarshal(payload, &c); err != nil {
		return "", fmt.Errorf("parse payload: %w", err)
	}
	if c.Iss != s.cfg.PublicURL {
		return "", errors.New("issuer mismatch")
	}
	if c.Aud != s.cfg.Audience {
		return "", errors.New("audience mismatch")
	}
	now := s.now().Unix()
	if c.Exp < now {
		return "", errors.New("expired")
	}
	// 5s leeway on iat to absorb clock drift between issuer and verifier
	// (here the same process — drift is zero by construction, but it keeps
	// the check symmetric with exp and makes the contract explicit).
	if c.Iat > now+5 {
		return "", errors.New("issued in the future")
	}
	return c.Sub, nil
}

// pkceVerify confirms that SHA256(verifier) base64url-encoded equals challenge.
// Method must be "S256"; "plain" is rejected per OAuth 2.1.
func pkceVerify(verifier, challenge, method string) error {
	if method != pkceMethodS256 {
		return fmt.Errorf("unsupported code_challenge_method: %s", method)
	}
	sum := sha256.Sum256([]byte(verifier))
	got := base64.RawURLEncoding.EncodeToString(sum[:])
	if !hmac.Equal([]byte(got), []byte(challenge)) {
		return errors.New("PKCE verifier does not match challenge")
	}
	return nil
}

File: src/oauth/jwt.rs
//! HS256 JWT issue/verify + PKCE S256 verification. Port of Go's
//! `internal/oauth/jwt.go`. Symmetric keys are fine: the same process issues
//! and validates, so there's no separate verifier needing public-key crypto.

use std::sync::LazyLock;

use base64::Engine;
use base64::engine::general_purpose::URL_SAFE_NO_PAD as B64URL;
use hmac::{Hmac, KeyInit, Mac};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use subtle::ConstantTimeEq;

type HmacSha256 = Hmac<Sha256>;

/// Fixed HS256 header; its base64url form is precomputed once.
const JWT_HEADER: &str = r#"{"alg":"HS256","typ":"JWT"}"#;
static JWT_HEADER_B64: LazyLock<String> = LazyLock::new(|| B64URL.encode(JWT_HEADER));

/// Access-token payload: standard registered claims + a free-form `scope`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Claims {
    pub iss: String,
    pub sub: String,
    pub aud: String,
    pub iat: i64,
    pub exp: i64,
    pub jti: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub scope: String,
}

#[derive(Debug, PartialEq, Eq)]
pub enum JwtError {
    Malformed,
    BadSignature,
    DecodePayload,
    ParsePayload,
    IssuerMismatch,
    AudienceMismatch,
    Expired,
    IssuedInFuture,
}

impl JwtError {
    /// Short token for the WWW-Authenticate error_description (not for debug logs).
    pub fn as_str(&self) -> &'static str {
        match self {
            JwtError::Malformed => "malformed JWT",
            JwtError::BadSignature => "bad signature",
            JwtError::DecodePayload => "decode payload",
            JwtError::ParsePayload => "parse payload",
            JwtError::IssuerMismatch => "issuer mismatch",
            JwtError::AudienceMismatch => "audience mismatch",
            JwtError::Expired => "expired",
            JwtError::IssuedInFuture => "issued in the future",
        }
    }
}

/// Signs `claims` with the HS256 secret, returning the compact JWT.
pub fn sign(secret: &[u8], claims: &Claims) -> Result<String, String> {
    let payload = serde_json::to_vec(claims).map_err(|e| format!("marshal claims: {e}"))?;
    let payload_b64 = B64URL.encode(payload);
    let signing = format!("{}.{}", *JWT_HEADER_B64, payload_b64);
    let sig_b64 = B64URL.encode(hmac_sha256(secret, signing.as_bytes()));
    Ok(format!("{signing}.{sig_b64}"))
}

/// Parses, signature-verifies and claim-checks a token issued by this server.
/// Returns the validated claims. `now` is unix seconds.
pub fn verify(
    secret: &[u8],
    token: &str,
    iss: &str,
    aud: &str,
    now: i64,
) -> Result<Claims, JwtError> {
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return Err(JwtError::Malformed);
    }
    let (header_b64, payload_b64, sig_b64) = (parts[0], parts[1], parts[2]);

    // Constant-time signature compare.
    let signing = format!("{header_b64}.{payload_b64}");
    let expected = hmac_sha256(secret, signing.as_bytes());
    let provided = B64URL.decode(sig_b64).map_err(|_| JwtError::BadSignature)?;
    if expected.as_slice().ct_eq(provided.as_slice()).unwrap_u8() != 1 {
        return Err(JwtError::BadSignature);
    }

    let payload = B64URL
        .decode(payload_b64)
        .map_err(|_| JwtError::DecodePayload)?;
    let claims: Claims = serde_json::from_slice(&payload).map_err(|_| JwtError::ParsePayload)?;

    if claims.iss != iss {
        return Err(JwtError::IssuerMismatch);
    }
    if claims.aud != aud {
        return Err(JwtError::AudienceMismatch);
    }
    // RFC 7519 §4.1.4: a token is expired when now >= exp.
    if claims.exp <= now {
        return Err(JwtError::Expired);
    }
    // 5s leeway on iat for clock drift (zero here by construction, but explicit).
    if claims.iat > now + 5 {
        return Err(JwtError::IssuedInFuture);
    }
    Ok(claims)
}

/// Confirms base64url(SHA-256(verifier)) == challenge. Method must be "S256".
pub fn pkce_verify(verifier: &str, challenge: &str, method: &str) -> Result<(), String> {
    if method != "S256" {
        return Err(format!("unsupported code_challenge_method: {method}"));
    }
    let got = B64URL.encode(Sha256::digest(verifier.as_bytes()));
    if got.as_bytes().ct_eq(challenge.as_bytes()).unwrap_u8() != 1 {
        return Err("PKCE verifier does not match challenge".to_string());
    }
    Ok(())
}

fn hmac_sha256(secret: &[u8], data: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(secret).expect("HMAC accepts any key length");
    mac.update(data);
    mac.finalize().into_bytes().to_vec()
}

#[cfg(test)]
mod tests {
    use super::*;

    const SECRET: &[u8] = b"test-secret-at-least-32-bytes-long!!";

    fn claims(now: i64) -> Claims {
        Claims {
            iss: "https://mcp.example".into(),
            sub: "dusan".into(),
            aud: "https://mcp.example/mcp".into(),
            iat: now,
            exp: now + 3600,
            jti: "jti1".into(),
            scope: "openid offline_access".into(),
        }
    }

    #[test]
    fn sign_verify_roundtrip() {
        let tok = sign(SECRET, &claims(1000)).unwrap();
        let got = verify(
            SECRET,
            &tok,
            "https://mcp.example",
            "https://mcp.example/mcp",
            2000,
        )
        .unwrap();
        assert_eq!(got.sub, "dusan");
        assert_eq!(got.scope, "openid offline_access");
    }

    #[test]
    fn tampered_signature_rejected() {
        let mut tok = sign(SECRET, &claims(1000)).unwrap();
        tok.pop();
        tok.push(if tok.ends_with('A') { 'B' } else { 'A' });
        assert_eq!(
            verify(
                SECRET,
                &tok,
                "https://mcp.example",
                "https://mcp.example/mcp",
                2000
            ),
            Err(JwtError::BadSignature)
        );
    }

    #[test]
    fn wrong_secret_rejected() {
        let tok = sign(SECRET, &claims(1000)).unwrap();
        assert_eq!(
            verify(
                b"another-secret-also-32-bytes-long!!!",
                &tok,
                "https://mcp.example",
                "https://mcp.example/mcp",
                2000
            ),
            Err(JwtError::BadSignature)
        );
    }

    #[test]
    fn audience_and_issuer_checked() {
        let tok = sign(SECRET, &claims(1000)).unwrap();
        assert_eq!(
            verify(
                SECRET,
                &tok,
                "https://mcp.example",
                "https://wrong/mcp",
                2000
            ),
            Err(JwtError::AudienceMismatch)
        );
        assert_eq!(
            verify(
                SECRET,
                &tok,
                "https://wrong",
                "https://mcp.example/mcp",
                2000
            ),
            Err(JwtError::IssuerMismatch)
        );
    }

    #[test]
    fn expiry_and_future_checked() {
        let tok = sign(SECRET, &claims(1000)).unwrap();
        // exp == now → expired.
        assert_eq!(
            verify(
                SECRET,
                &tok,
                "https://mcp.example",
                "https://mcp.example/mcp",
                4600
            ),
            Err(JwtError::Expired)
        );
        // iat far in the future.
        let future = sign(SECRET, &claims(10_000)).unwrap();
        assert_eq!(
            verify(
                SECRET,
                &future,
                "https://mcp.example",
                "https://mcp.example/mcp",
                100
            ),
            Err(JwtError::IssuedInFuture)
        );
    }

    #[test]
    fn pkce_s256_roundtrip() {
        // verifier → challenge = base64url(sha256(verifier))
        let verifier = "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk";
        let challenge = B64URL.encode(Sha256::digest(verifier.as_bytes()));
        assert!(pkce_verify(verifier, &challenge, "S256").is_ok());
        assert!(pkce_verify("wrong-verifier", &challenge, "S256").is_err());
        assert!(pkce_verify(verifier, &challenge, "plain").is_err());
    }
}
